crates/iota-indexer/src/config.rs

```diff
@@ -0,0 +1,556 @@
+            registry_id,
+            reverse_registry_id,
+        } = config;
+        Self {
+            package_address,
+            object_id,
+            payments_package_address,
+            registry_id,
+            reverse_registry_id,
+        }
+    }
+}
+
+impl Default for IotaNamesOptions {
+    fn default() -> Self {
+        IotaNamesConfig::default().into()
+    }
+}
+
+#[derive(Args, Debug, Clone)]
+pub struct JsonRpcConfig {
+    #[command(flatten)]
+    pub iota_names_options: IotaNamesOptions,
+
+    #[clap(long, default_value = "0.0.0.0:9000")]
+    pub rpc_address: SocketAddr,
+
+    #[clap(long)]
+    pub rpc_client_url: String,
+}
+
+#[derive(Args, Debug, Default, Clone)]
+#[group(required = true, multiple = true)]
+pub struct IngestionSources {
+    #[arg(long)]
+    pub data_ingestion_path: Option<PathBuf>,
+
+    #[arg(long)]
+    pub remote_store_url: Option<Url>,
+
+    #[arg(long)]
+    pub rpc_client_url: Option<Url>,
+}
+
+#[derive(Args, Debug, Clone)]
+pub struct IngestionConfig {
+    #[clap(flatten)]
+    pub sources: IngestionSources,
+
+    #[arg(
+        long,
+        default_value_t = Self::DEFAULT_CHECKPOINT_DOWNLOAD_QUEUE_SIZE,
+        env = "DOWNLOAD_QUEUE_SIZE",
+    )]
+    pub checkpoint_download_queue_size: usize,
+
+    #[arg(
+        long,
+        default_value_t = Self::DEFAULT_CHECKPOINT_DOWNLOAD_TIMEOUT,
+        env = "INGESTION_READER_TIMEOUT_SECS",
+    )]
+    pub checkpoint_download_timeout: u64,
+
+    /// Limit indexing parallelism on big checkpoints to avoid OOMing by
+    /// limiting the total size of the checkpoint download queue.
+    #[arg(
+        long,
+        default_value_t = Self::DEFAULT_CHECKPOINT_DOWNLOAD_QUEUE_SIZE_BYTES,
+        env = "CHECKPOINT_PROCESSING_BATCH_DATA_LIMIT",
+    )]
+    pub checkpoint_download_queue_size_bytes: usize,
+}
+
+impl IngestionConfig {
+    pub const DEFAULT_CHECKPOINT_DOWNLOAD_QUEUE_SIZE: usize = 200;
+    pub const DEFAULT_CHECKPOINT_DOWNLOAD_QUEUE_SIZE_BYTES: usize = 20_000_000;
+    pub const DEFAULT_CHECKPOINT_DOWNLOAD_TIMEOUT: u64 = 20;
+}
+
+impl Default for IngestionConfig {
+    fn default() -> Self {
+        Self {
+            sources: Default::default(),
+            checkpoint_download_queue_size: Self::DEFAULT_CHECKPOINT_DOWNLOAD_QUEUE_SIZE,
+            checkpoint_download_timeout: Self::DEFAULT_CHECKPOINT_DOWNLOAD_TIMEOUT,
+            checkpoint_download_queue_size_bytes:
+                Self::DEFAULT_CHECKPOINT_DOWNLOAD_QUEUE_SIZE_BYTES,
+        }
+    }
+}
+
+#[derive(Subcommand, Clone, Debug)]
+pub enum Command {
+    Indexer {
+        #[command(flatten)]
+        ingestion_config: IngestionConfig,
+        #[command(flatten)]
+        snapshot_config: SnapshotLagConfig,
+        #[command(flatten)]
+        pruning_options: PruningOptions,
+        #[arg(long)]
+        reset_db: bool,
+    },
+    JsonRpcService(JsonRpcConfig),
+    AnalyticalWorker,
+    HelpDeprecated,
```

Comment:
Nit: Add docstring to show while calling `--help`
```suggestion
    /// Print help for the deprecated interface.
    HelpDeprecated,
```